### DISCUSSION TITLE
Type-issue using "set" in a custom TypeScript middleware

### DISCUSSION THREAD

**Discussion by edolix** [pmndrs/zustand#1780] [Q&A] [answered]
## Summary

I'm trying to create a middleware using TypeScript to automate some tasks. I'm facing a **type-only** issue when using `set({ ... })` inside the middleware. 
The problem seems related to TS not able to infer the type of the `set()` argument. 

In the [CodeSandbox](https://codesandbox.io/s/relaxed-hellman-ih8qs9?file=/src/store.ts:1116-1176) you can see that i can't use `set({ bears: nextState.bears });` or `set({ bears: 2 });` because TS complains with:

```
Argument of type '{ bears: T["bears"] | undefined; }' is not assignable to parameter of type 'T | Partial<T> | ((state: T) => T | Partial<T>)'.
//
Argument of type '{ bears: 2; }' is not assignable to parameter of type 'T | Partial<T> | ((state: T) => T | Partial<T>)'.
```

I don't understand _why_ `{ bears: <number> }` is not assignable so maybe something that can be improved in the Zustand type-system?

**Note:** `exactOptionalPropertyTypes` is [enabled in the CodeSandbox](https://codesandbox.io/s/relaxed-hellman-ih8qs9?file=/tsconfig.json:129-173).

Might be related to:
 
  - https://github.com/pmndrs/zustand/issues/1723
  - https://github.com/pmndrs/zustand/discussions/1770#discussioncomment-5726475

Thank you so much for this library 🙌 

## Link to reproduction

See the [store.ts in this CodeSandbox](https://codesandbox.io/s/relaxed-hellman-ih8qs9?file=/src/store.ts:1116-1153)

## Check List

Please do not ask questions in issues.

- [ ] I've already opened a [discussion](https://github.com/pmndrs/zustand/discussions) before opening this issue, or already discussed in other media.

Please include a minimal reproduction.

- [x] I've added a link to a [codesandbox](https://codesandbox.io/s/relaxed-hellman-ih8qs9?file=/src/store.ts:1116-1153) with a minimal reproduction.

Please check this if you're filing an issue regarding TypeScript.

- [x] I've read the [typescript guide](https://github.com/pmndrs/zustand/blob/main/docs/guides/typescript.md), in particular that `create` is to be used as `create<T>()(...)` and not `create<T>(...)`.

**Comment by dai-shi** [comment, marked as answer]
This is what I tried: https://tsplay.dev/mqYLZm
Not 100% sure if it's the right one and if there's a better one.

**Reply by edolix** [reply]
Thanks for the quick reply @dai-shi ! 

Looks like you just removed the generic from the `type LoggerImpl`. Not sure if the _real_ problem is within the Zustand types not working well with the generic but this change did the trick! 
Thank you 🎉 


See the [working solution here](https://codesandbox.io/s/relaxed-hellman-ih8qs9?file=/src/store.ts:975-1115)
